compiler/rustc_middle/src/ty/layout_sanity_check.rs

```diff
@@ -0,0 +1,303 @@
+use crate::ty::{
+    layout::{LayoutCx, TyAndLayout},
+    TyCtxt,
+};
+use rustc_target::abi::*;
+
+use std::cmp;
+
+/// Enforce some basic invariants on layouts.
+pub(super) fn sanity_check_layout<'tcx>(
+    cx: &LayoutCx<'tcx, TyCtxt<'tcx>>,
+    layout: &TyAndLayout<'tcx>,
+) {
+    // Type-level uninhabitedness should always imply ABI uninhabitedness.
+    if cx.tcx.conservative_is_privately_uninhabited(cx.param_env.and(layout.ty)) {
+        assert!(layout.abi.is_uninhabited());
+    }
+
+    if layout.size.bytes() % layout.align.abi.bytes() != 0 {
+        bug!("size is not a multiple of align, in the following layout:\n{layout:#?}");
+    }
+
+    if cfg!(debug_assertions) {
+        /// Yields non-1-ZST fields of the type
+        fn non_zst_fields<'tcx, 'a>(
```

Comment:
Could this have `1zst`/`1_zst` or `align1zst`/`align1_zst` in the name instead of `zst`?

Reply:
It doesn't actually produce 1-ZST though, I had to change it to produce all ZST. Will adjust the comment.

Reply:
FWIW this function seems useful to have elsewhere as well? Not sure where to put it though.